src/Interactive/Features/Interactive/Core/InteractiveHost.RemoteService.cs

```diff
@@ -49,35 +57,32 @@ internal RemoteService(InteractiveHost host, Process process, int processId, Ser
 
             internal void HookAutoRestartEvent()
             {
-                const int ProcessExitHooked = 1;
-                const int ProcessExitHandled = 2;
-
-                int processExitHandling = 0;
+                // hook the event only once per process:
+                if (Interlocked.Exchange(ref _processExitHandling, ProcessExitHooked) == 0)
+                {
+                    Process.Exited += ProcessExitedHandler;
+                }
+            }
 
-                async void localHandler(object _, EventArgs __)
+            private async void ProcessExitedHandler(object _, EventArgs __)
+            {
+                try
                 {
-                    try
+                    if (Interlocked.Exchange(ref _processExitHandling, ProcessExitHandled) == ProcessExitHooked)
                     {
-                        if (Interlocked.Exchange(ref processExitHandling, ProcessExitHandled) == ProcessExitHooked)
-                        {
-                            Process.Exited -= localHandler;
+                        Process.Exited -= ProcessExitedHandler;
 
-                            if (!_disposing)
-                            {
-                                await _host.OnProcessExited(Process).ConfigureAwait(false);
-                            }
+                        if (!_disposing && _host != null)
```

Comment:
>_host != null [](start = 43, length = 13)

Good catch here, but there is still race. ```_host?.OnProcessExited(...)``` would be better.

Reply:
Actually, since _disposing must be true before _host is set to null, _host can't be null at this point. 

Make _disposing volatile field though, so that the JIT doesn't reorder.

---
In reply to: [162224558](https://github.com/dotnet/roslyn/pull/24058#discussion_r162224558) [](ancestors = 162224558)